create-cluster.sh

```diff
@@ -0,0 +1,179 @@
+readonly DNS_FLUSH_COMMAND='sudo killall -HUP mDNSResponder' # Mac OSX Mojave
+
+set -euo pipefail
+
+main() {
+  local readonly cluster_name=$1
+
+  check_prerequisites ${cluster_name}
+
+  git-crypt unlock
+  get_sudo
+
+  create_cluster ${cluster_name}
+  run_kops ${cluster_name}
+  install_components ${cluster_name}
+
+  kubectl cluster-info
+}
+
+create_cluster() {
+  local readonly cluster_name=$1
+  (
+    cd terraform/cloud-platform
+    switch_terraform_workspace ${cluster_name}
+    terraform apply -auto-approve
+  )
+}
+
+run_kops() {
+  local readonly cluster_name=$1
+
+  kops create -f kops/${cluster_name}.yaml
+
+  # This is a throwaway SSH key which we never need again.
+  rm -rf /tmp/${cluster_name}*
+  ssh-keygen -b 4096 -P '' -f /tmp/${cluster_name}
+
+  kops create secret --name ${cluster_name}.${CLUSTER_SUFFIX} sshpublickey admin -i /tmp/${cluster_name}.pub
+  kops update cluster ${cluster_name}.${CLUSTER_SUFFIX} --yes --alsologtostderr
+
+  wait_for_kops_validate
+}
+
+# TODO: figure out this problem, and fix it.
+# For some reason, the first terraform apply sometimes fails with an error "could not find a ready tiller pod"
+# This seems to be quite misleading, since adding a delay after 'helm init' makes no difference.
+# A second run of the terraform apply usually works correctly.
+install_components() {
+  local readonly cluster_name=$1
+  cd terraform/cloud-platform-components
+  switch_terraform_workspace ${cluster_name}
+
+  if terraform apply -auto-approve
+  then
+    echo "Cluster components installed."
+  else
+    echo "Initial components install reported errors. Sleeping and retrying..."
+    sleep 120
+    terraform apply -auto-approve
+  fi
+}
+
+wait_for_kops_validate() {
+  local readonly max_tries=30
+  local validated=0
+
+  for attempt in $(seq 1 $max_tries); do
+    echo "Validate cluster, attempt ${attempt} of $max_tries..."
+
+    if kops validate cluster
+    then
+      echo "Cluster validated."
+      validated=1
+      break
+    else
+      echo "Flushing DNS and sleeping before retry..."
+      ${DNS_FLUSH_COMMAND}
+      sleep 60
+    fi
+  done
+
+  if [ "${validated}" -eq 0 ]; then
+    echo "Failed to validate cluster after $max_tries attempts."
+    exit 1
+  fi
+}
+
+switch_terraform_workspace() {
+  local readonly name=$1
+  terraform init
+  # The workspace might already exist, so the workspace new is allowed to fail
+  # but the workspace select must succeed
+  terraform workspace new ${name} || true
+  terraform workspace select ${name}
+}
+
+check_prerequisites() {
+  local readonly cluster_name=$1
+
+  check_env_vars
+  check_software_installed
+  check_aws_profiles
+  check_name_length ${cluster_name}
+
+  # TODO: check helm version is >= 2.11
+}
+
+check_env_vars() {
+  test $AWS_PROFILE
+  test $AUTH0_DOMAIN
+  test $AUTH0_CLIENT_ID
+  test $AUTH0_CLIENT_SECRET
+  test $KOPS_STATE_STORE
+}
+
+check_software_installed() {
+  hash git-crypt
```

Comment:
Curious, why `hash` instead of `which`?

Reply:
exit status

https://stackoverflow.com/a/677212/794111

Reply:
Interesting, thanks for the link that's a treasure of an answer. 
FWIW, my `which` returns an exit code just fine and `hash` does too on `zsh`.

Reply:
I think it's one of those things that will almost always work, but the answer also states that hash is a built-in, and which spawns a separate process, so it's arguably more efficient, and theoretically more reliable. There's very little in it. I think I'll add that link as a comment, to pre-empt future confusion!